### PR TITLE
[POC]  Graceful shutdown

### PR DESCRIPTION
**Pull request by NYBACHOK**

### CODE REVIEW

File: tendermint/src/application.rs [hunk 1]
```diff
@@ -110,109 +115,200 @@ pub trait ABCIApplication<G>: Send + Clone + 'static {
...
         request: tendermint_proto::abci::RequestCheckTx,
     ) -> tendermint_proto::abci::ResponseCheckTx {
-        T::check_tx(&self.handler, request.into()).into()
+        match self.token.if_cancelled_ctx() {
+            Some(ctx) => match ctx {
+                CancellationContext::GasOverflow => {
+                    ResponseCheckTx::error_with_gas_overflow().into()
+                }
+                _ => {
+                    self.token.panic_if_cancelled();
```
Comment (dmvict) [inline tendermint/src/application.rs:201]: What are the other cases here? It's a little unclear here.

Comment (NYBACHOK) [reply tendermint/src/application.rs:201]: Whatever you like. This code only to show possible usage for gas and maybe something else

Comment (dmvict) [reply tendermint/src/application.rs:201]: > Whatever you like. This code only to show possible usage for gas and maybe something else

Okay. But improved version may have concrete variants except gas metering and it's better to add them.

Comment (NYBACHOK) [reply tendermint/src/application.rs:201]: I'm unsure that we need gas in token...

**Pull request closed without merging**